contracts/FeralfileEnglishAuction.sol

```diff
@@ -0,0 +1,317 @@
+        // get the auction object
+        // get the last auction bid object
+        // check if current winner not equal to current bidder (optional)
+        // check if now() under auction end time
+        // check bidding amount > highest bid amount
+        // check bidding amount follow the minimum increment
+        // check timestamp should in 10 mins
+
+        // check the signature of (aucId + bidder + amount + timestamp) to make sure the request is from Feral File
+        bytes32 message_ = keccak256(
+            abi.encode(block.chainid, aucId_, bidder_, amount_, timestamp_)
+        );
+        require(
+            isValidSignature(message_, r_, s_, v_),
+            "FeralfileEnglishAuction: invalid signature"
+        );
+        // transfer last winning bid amount to last bidder if fromFeralFile is false
+        if (!lastBid_.fromFeralFile) {
+            payable(lastBid_.bidder).transfer(lastBid_.amount);
+        }
+        // replace the new bid to bid map
+        auctionLatestBid[aucId_] = Bid({
+            bidder: bidder_,
+            amount: amount_,
+            fromFeralFile: true
+        });
+        // if the gap to end time is lower than threshold, update the auction end time to now() + extend duration
+        if (auction_.endAt - block.timestamp < auction_.extendThreshold) {
+            auctions[aucId_].endAt = block.timestamp + auction_.extendDuration;
+        }
+        // emit new bid event
+        emit NewBid(aucId_, bidder_, amount_);
+    }
+
+    function settleAuction(
+        bytes32 r_,
+        bytes32 s_,
+        uint8 v_,
+        address tokenAddr_,
+        address vaultAddr_,
+        SaleData memory saleData_
+    ) external onlyOwner {
+        // transfer winning bid amount to Feral File Vault contract if winning bid is crypto bid
+        bytes32 aucId_ = getAuctionID(tokenAddr_, saleData_.tokenIds);
+
+        EnglishAuction memory auction_ = auctions[aucId_];
+        require(auction_.id != 0, "FeralfileEnglishAuction: auction not found");
+
+        require(
+            auction_.endAt < block.timestamp,
+            "FeralfileEnglishAuction: auction not ended"
+        );
+
+        Bid memory lastBid_ = auctionLatestBid[aucId_];
+        require(
+            lastBid_.bidder != address(0),
+            "FeralfileEnglishAuction: no bid"
+        );
+
+        require(
+            !auction_.isSettled,
+            "FeralfileEnglishAuction: auction already settled"
+        );
+
+        auctions[aucId_].isSettled = true;
+
+        saleData_.price = lastBid_.amount;
+        saleData_.destination = lastBid_.bidder;
+        saleData_.payByVaultContract = true;
+
+        if (!lastBid_.fromFeralFile) {
+            payable(vaultAddr_).transfer(lastBid_.amount);
+        }
+        FFV4(tokenAddr_).buyArtworks(r_, s_, v_, saleData_);
+    }
+
+    event NewBid(
+        bytes32 indexed auctionId,
+        address indexed bidder,
+        uint256 indexed amount
```

Comment:
could you help to emit fromferalfile too?

Reply:
added